src/bp/core/services/converse.ts

```diff
@@ -0,0 +1,147 @@
+import { UserRepository } from 'core/repositories'
+import { TYPES } from 'core/types'
+import { InvalidParameterError } from 'errors'
+import { EventEmitter2 } from 'eventemitter2'
+import { inject, injectable, postConstruct } from 'inversify'
+import _ from 'lodash'
+
+import { Event } from '../sdk/impl'
+
+import { EventEngine } from './middleware/event-engine'
+
+export const converseApiEvents = new EventEmitter2()
+
+@injectable()
+export class ConverseService {
+  private readonly timeoutInMs = 2000
+  private jsonMap: Map<string, object> = new Map()
+
+  constructor(
+    @inject(TYPES.EventEngine) private eventEngine: EventEngine,
+    @inject(TYPES.UserRepository) private userRepository: UserRepository
+  ) {}
+
+  @postConstruct()
+  init() {
+    this.eventEngine.register({
+      name: 'converse.capture.payload',
+      description: 'Captures the response payload for the Converse API',
+      order: 10000,
+      direction: 'outgoing',
+      handler: (event, next) => {
+        this._handleCapturePayload(event)
+        next()
+      }
+    })
+    this.eventEngine.register({
+      name: 'converse.capture.nlu',
+      description: 'Captures the nlu output for the Converse API',
+      order: 10000,
+      direction: 'incoming',
+      handler: (event, next) => {
+        this._handleCaptureNlu(event)
+        next()
+      }
+    })
+  }
+
+  public async sendMessage(botId: string, userId: string, payload): Promise<any> {
+    if (!payload.text || !_.isString(payload.text) || payload.text.length > 360) {
+      throw new InvalidParameterError('Text must be a valid string of less than 360 chars')
+    }
+
+    await this.userRepository.getOrCreate('api', userId)
+
+    const incomingEvent = Event({
+      type: 'text',
+      channel: 'api',
+      direction: 'incoming',
+      payload,
+      target: userId,
+      botId
+    })
+
+    const timeoutPromise = this._createTimeoutPromise(userId)
+    const donePromise = this._createDonePromise(userId)
+
+    await this.eventEngine.sendEvent(incomingEvent)
+
+    return Promise.race([timeoutPromise, donePromise])
+  }
+
+  private async _createDonePromise(userId) {
+    return new Promise((resolve, reject) => {
+      const doneEvent = `done.${userId}`
+
+      converseApiEvents.removeAllListeners(doneEvent)
```

Comment:
why here? it's never going to do anything. you must call that line 69 in a finally